cdap-ui/app/cdap/components/FastAction/SetPreferenceAction/SetPreferenceModal.js

```diff
@@ -0,0 +1,430 @@
+    let keyValObj = {};
+    keyValArr.forEach((pair) => {
+      if (pair.key.length > 0 && pair.value.length > 0) {
+        keyValObj[pair.key] = pair.value;
+      }
+    });
+    return keyValObj;
+  }
+
+  toggleSorted(attribute) {
+    let sortOrder = 'asc';
+    if (this.state.sortByAttribute != attribute) {
+      this.setState({sortOrder});
+    } else {
+      if (this.state.sortOrder === 'asc') {
+        sortOrder = 'desc';
+      }
+      this.setState({sortOrder});
+    }
+    this.setState({sortByAttribute: attribute});
+    let newInheritedPreferences = orderBy(this.state.inheritedPreferences, attribute, sortOrder);
+    this.setState({inheritedPreferences: newInheritedPreferences});
+  }
+
+  onKeyValueChange(keyValues) {
+    if (!this.state.fieldsResetted) {
+      this.setState({keyValues});
+    }
+    else {
+      this.setState({fieldsResetted: false});
+    }
+  }
+
+  allFieldsFilled() {
+    if (this.state.keyValues.pairs) {
+      return this.state.keyValues.pairs.every((keyValuePair) => {
+        return (keyValuePair.key.length > 0 && keyValuePair.value.length > 0);
+      });
+    }
+    return true;
+  }
+
+  resetFields(event) {
+    event.preventDefault();
+    this.getSpecifiedPreferences();
+  }
+
+  preventPropagation(event) {
+    event.stopPropagation();
+    event.nativeEvent.stopImmediatePropagation();
+  }
+
+  componentWillMount() {
+    this.getSpecifiedPreferences();
+    this.getInheritedPreferences();
+  }
+
+  renderSpecifyPreferences() {
+    const actionLabel = T.translate('features.FastAction.setPreferencesActionLabel');
+    let entity = "Namespace " + this.params.namespace;
+    if (this.props.entity) {
+      entity = upperFirst(this.props.entity.type) + " " + this.props.entity.id;
+    }
+    const title = `${actionLabel} for ${entity}`;
+    const keyLabel = T.translate('features.FastAction.setPreferencesColumnLabel.key');
+    const valueLabel = T.translate('features.FastAction.setPreferencesColumnLabel.value');
+    let description = T.translate('features.FastAction.setPreferencesDescriptionLabel.namespace');
+    if (this.props.entity) {
+      if (this.props.entity.type == 'application') {
+        description = T.translate('features.FastAction.setPreferencesDescriptionLabel.app');
+      } else {
+        description = T.translate('features.FastAction.setPreferencesDescriptionLabel.program');
+      }
+    }
+    return (
+      <div>
+        <div className='specify-preferences-description'>
+          <h4>{title}</h4>
+          <p>{description}</p>
+        </div>
+        <div className='specify-preferences-list'>
+          <div className='specify-preferences-labels'>
+            <span className='key-label'>{keyLabel}</span>
+            <span className='value-label'>{valueLabel}</span>
+          </div>
+          <div className='specify-preferences-values'>
+            <KeyValuePairs
+              keyValues = {this.state.keyValues}
```

Comment:
if `keyValues` is an empty error there is a js error thrown in the console 
<img width="1100" alt="screen shot 2017-01-23 at 9 29 23 pm" src="https://cloud.githubusercontent.com/assets/1452845/22235267/5c6e2588-e1b3-11e6-8402-f0ea827560fb.png">